core/rovecomm.py

```diff
@@ -141,46 +277,187 @@ def read(self):
+                data = packet[header_size:]
 
-            if rovecomm_version != 2:
-                return_packet = RoveCommPacket(ROVECOMM_INCOMPATIBLE_VERSION, 'b', (1,), '')
+                if rovecomm_version != 2:
+                    return_packet = RoveCommPacket(ROVECOMM_INCOMPATIBLE_VERSION, 'b', (1,), '')
+                    return_packet.ip_address = remote_ip
+                    return return_packet
+
+                if data_id == ROVECOMM_SUBSCRIBE_REQUEST:
+                    if self.subscribers.count(remote_ip) == 0:
+                        self.subscribers.append(remote_ip)
+                elif data_id == ROVECOMM_UNSUBSCRIBE_REQUEST:
+                    if self.subscribers.count(remote_ip) != 0:
+                        self.subscribers.remove(remote_ip)
+
+                data_type = types_int_to_byte[data_type]
+                data = struct.unpack('>' + data_type * data_count, data)
+
+                return_packet = RoveCommPacket(data_id, data_type, data, '')
                 return_packet.ip_address = remote_ip
                 return return_packet
 
-            if data_id == ROVECOMM_SUBSCRIBE_REQUEST:
-                if self.subscribers.count(remote_ip == 0):
-                    self.subscribers.append(remote_ip)
-            elif data_id == ROVECOMM_UNSUBSCRIBE_REQUEST:
-                if self.subscribers.count(remote_ip) != 0:
-                    self.subscribers.remove(remote_ip)
+            except Exception:
+                return_packet = RoveCommPacket()
+                return return_packet
+
+    def close_socket(self):
+        '''
+        Closes the UDP socket
+        '''
+        self.RoveCommSocket.shutdown(1)
+        self.RoveCommSocket.close()
 
-            data_type = types_int_to_byte[data_type]
-            data = struct.unpack('>' + data_type * data_count, data)
 
-            return_packet = RoveCommPacket(data_id, data_type, data, '')
-            return_packet.ip_address = remote_ip
-            return return_packet
+class RoveCommEthernetTcp:
+    '''
+    The TCP implementation for RoveComm.
+
+    Methods:
+        write(packet):
+            Transmits a packet to the destination IP and all active subscribers.
+        read():
+            Receives all TCP packets from open sockets and packs data into RoveCommPacket instances
+        connect(ip_octet):
+            Opens a socket connection to the given address
+        close_sockets():
+            Closes the server socket and all open sockets
+        handle_incoming_connections():
+            Accepts socket connection requests
+    '''
+    def __init__(self, HOST=socket.gethostbyname(socket.gethostname()), PORT=ROVECOMM_TCP_PORT):
+        self.open_sockets = {}
+        # configure a TCP socket
+        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        # bind the socket to the current machines local network IP by default (can be specified as well)
```

Comment:
should add the following line: self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1). This will allow us to reopen the socket instantly after we close it, allowing us to restart Autonomy code immediately.

Reply:
Add this before the bind() call